File: lib/aura/memory/recorder.rb
# frozen_string_literal: true

module Aura
  module Memory
    class Recorder
      def initialize(store)
        @store = store
      end

      def record_user(content, call_seq: nil)
        payload = { content: content.to_s, call_seq: call_seq }
        @store.insert_event(
          timestamp: Time.now.to_i,
          phase: "user",
          tool: nil,
          payload: payload.merge(phase: "user")
        )
      end

      def record_plan(plan)
        return nil unless plan.is_a?(Hash)

        plan_data = {
          tool: plan[:tool] || plan["tool"],
          args: plan[:args] || plan["args"] || {},
          summary: plan[:summary] || plan["summary"],
          thought: plan[:thought] || plan["thought"]
        }

        plan.each do |key, value|
          key_str = key.to_s
          next if %w[tool args summary thought type].include?(key_str)

          plan_data[key_str] = value
        end

        @store.insert_event(
          timestamp: Time.now.to_i,
          phase: "plan",
          tool: plan_data[:tool],
          payload: plan_data.merge(phase: "plan", tool: plan_data[:tool])
        )
      end

      def record_execution(tool_name, result, call_seq: nil)
        result_payload = result.is_a?(Hash) ? result : { output: result.to_s }
        payload = { result: result_payload, call_seq: call_seq }
        @store.insert_event(
          timestamp: Time.now.to_i,
          phase: "execution",
          tool: tool_name.to_s,
          payload: payload.merge(phase: "execution", tool: tool_name.to_s)
        )
      end

      def record_interception(tool_name, advice, reason: nil)
        payload = { advice: advice.to_s }
        payload[:reason] = reason if reason

        @store.insert_event(
          timestamp: Time.now.to_i,
          phase: "interception",
          tool: tool_name.to_s,
          payload: payload.merge(phase: "interception", tool: tool_name.to_s)
        )
      end

      def record_custom(phase, payload = {})
        tool = payload[:tool] || payload["tool"]
        @store.insert_event(
          timestamp: Time.now.to_i,
          phase: phase.to_s,
          tool: tool,
          payload: payload.merge(phase: phase.to_s, tool: tool)
        )
      end

      def record_summary(content, source_event_id = nil)
        @store.insert_summary(
          content: content,
          source_event_id: source_event_id
        )
      end

      def record_batch(events)
        @store.transaction do
          events.map do |event|
            type = event[:type] || event["type"]
            case type
            when "user"
              record_user(event[:content] || event["content"], call_seq: event[:call_seq])
            when "plan"
              record_plan(event[:plan] || event["plan_data"])
            when "execution"
              record_execution(
                event[:tool] || event["tool_name"],
                event[:result],
                call_seq: event[:call_seq]
              )
            when "interception"
              record_interception(
                event[:tool] || event["tool_name"],
                event[:advice],
                reason: event[:reason]
              )
            else
              record_custom(type, event.reject { |k, _| [:type].include?(k) })
            end
          end
        end
      end
    end
  end
end

File: src/core/memory/recorder.ts
import type { ParseResult } from '../llm/parsers/responseParser.js';
import type { SQLiteStore } from './sqliteStore.js';

export class MemoryRecorder {
  private store: SQLiteStore;

  constructor(store: SQLiteStore) {
    this.store = store;
  }

  public recordUser(content: string, callSeq?: number | null): number {
    const payload = { content, call_seq: callSeq, phase: 'user' };
    return this.store.insertEvent({
      timestamp: Math.floor(Date.now() / 1000),
      phase: 'user',
      tool: '',
      payload,
    });
  }

  public recordPlan(plan: ParseResult): number | null {
    if (!plan || typeof plan !== 'object') {
      return null;
    }

    const planData: Record<string, unknown> = {
      tool: plan.type === 'tool_call' ? plan.tool : null,
      args: plan.type === 'tool_call' ? plan.args : {},
      summary: plan.type === 'tool_call' ? plan.summary : null,
      thought: plan.thought ?? (plan.type === 'text' ? plan.content : null),
      phase: 'plan',
    };

    // Copy any extra attributes
    for (const [key, value] of Object.entries(plan)) {
      if (
        !['tool', 'args', 'summary', 'thought', 'type', 'phase'].includes(key)
      ) {
        planData[key] = value;
      }
    }

    return this.store.insertEvent({
      timestamp: Math.floor(Date.now() / 1000),
      phase: 'plan',
      tool: planData.tool as string,
      payload: planData,
    });
  }

  public recordExecution(
    toolName: string,
    result: Record<string, unknown>,
    callSeq?: number | null,
  ): number {
    const resultPayload =
      result && typeof result === 'object' && !Array.isArray(result)
        ? result
        : { output: String(result) };
    const payload = {
      result: resultPayload,
      call_seq: callSeq,
      phase: 'execution',
      tool: toolName,
    };
    return this.store.insertEvent({
      timestamp: Math.floor(Date.now() / 1000),
      phase: 'execution',
      tool: toolName,
      payload,
    });
  }

  public recordInterception(
    toolName: string,
    advice: string,
    reason?: string | null,
  ): number {
    const payload: Record<string, unknown> = {
      advice,
      phase: 'interception',
      tool: toolName,
    };
    if (reason) {
      payload.reason = reason;
    }
    return this.store.insertEvent({
      timestamp: Math.floor(Date.now() / 1000),
      phase: 'interception',
      tool: toolName,
      payload,
    });
  }

  public recordCustom(
    phase: string,
    payload: Record<string, unknown> = {},
  ): number {
    const tool = payload.tool ?? null;
    const cleanPayload = { ...payload, phase, tool };
    return this.store.insertEvent({
      timestamp: Math.floor(Date.now() / 1000),
      phase,
      tool: tool as string,
      payload: cleanPayload,
    });
  }

  public recordSummary(content: string, sourceEventId?: number | null): number {
    return this.store.insertSummary({
      content,
      source_event_id: sourceEventId ?? null,
    });
  }

  public recordBatch(events: Record<string, unknown>[]): void {
    this.store.transaction(() => {
      for (const event of events) {
        const type = event.type ?? event.phase;
        switch (type) {
          case 'user':
            this.recordUser(event.content as string, event.call_seq as number);
            break;
          case 'plan':
            this.recordPlan(
              (event.plan ?? event.plan_data ?? event) as ParseResult,
            );
            break;
          case 'execution':
            this.recordExecution(
              String(event.tool ?? event.tool_name ?? ''),
              (event.result as Record<string, unknown>) || {},
              event.call_seq as number,
            );
            break;
          case 'interception':
            this.recordInterception(
              String(event.tool ?? event.tool_name ?? ''),
              String(event.advice ?? ''),
              String(event.reason ?? ''),
            );
            break;
          default: {
            const { type: _, ...rest } = event;
            this.recordCustom(type as string, rest);
            break;
          }
        }
      }
    });
  }
}
